File: easy/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> ans;
        vector<int> pcount(26, 0);
        vector<int> scount(26, 0);

        for (int i=0; i<p.length(); i++) {
            scount['z' - s[i]] += 1;
            pcount['z' - p[i]] += 1;
        }

        bool isequal = true;

        for (int i=0; i<26; i++) {
            if (scount[i] != pcount[i]) {
                isequal = false;
                break;
            }
        }

        if (isequal) {
            ans.push_back(0);
        }

        int firstIndex = 0;

        for (int i = p.length(); i<s.length(); i++) {
            scount['z' - s[firstIndex++]] -= 1;
            scount['z' - s[i]] += 1;

            bool isequal = true;

            for (int j = 0; j<26; j++) {
                if (scount[j] != pcount[j]) {
                    isequal = false;
                    break;
                }
            }

            if (isequal) {
                ans.push_back(firstIndex);
            }
        }
        return ans;
    }
};

File: easy/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """

        ans = []

        plen = len(p)
        slen = len(s)
        
        if slen < plen:
            return ans

        sCounter = collections.Counter()
        pCounter = collections.Counter()
        
        for c in p:
            pCounter[c] += 1

        for i in range(plen):
            sCounter[s[i]] += 1

        firstIndex = 0

        if sCounter == pCounter:
                ans.append(firstIndex)

        for i in range(plen, slen):
            sCounter[s[firstIndex]] -= 1
            if sCounter[s[firstIndex]] == 0:
                del sCounter[s[firstIndex]]
            sCounter[s[i]] += 1
            firstIndex += 1
            if sCounter == pCounter:
                ans.append(firstIndex)

        return ans
